### ISSUE TITLE
Add ASIO/WASAPI drivers to Reduce Audio Latency

### ISSUE THREAD

**Issue by posesix** [ppy/osu-framework#705]
Opening an issue for discussion since this is something that should be included in the official distribution.

osu! uses default BASS procedures, which use DirectSound on Windows, causing large latencies due to high latency software mixing in Windows Vista and Windows 7 (as I've tested). I get **70ms latency** on average using 240FPS on both my Windows 7 64-bit PC and Windows Vista 32-bit laptop. This may be less of a problem on Linux since BASS defaults to the ALSA driver, which is a low-level & low-latency driver, but I have not tested it fully.

The BASS API has extensions for ASIO and WASAPI support under the same licenses.
We can effectively (hopefully) reduce the latency to less than 4ms on ASIO, and a worse case on WASAPI (Exclusive).

BASS does not unify their main API with ASIO and WASAPI APIs for some reason, so some changes and additions to audio code will be needed. These APIs are very easy to work with, and pretty much the same as using their main API once you've got them initialised. We have to set up a BASS MIXER and load sounds and music as BASS STREAMS, instead of SAMPLES. The STREAMS are connected to the MIXER and the MIXER feeds audio data to the ASIO/WASAPI event procedures. BASS CHANNEL procedures can be used with streams just like samples.

We will also need to add options for sound drivers in the options menu. It will simply list available drivers depending on your platform. A simple drop-down box can be created after detecting available drivers. Options could be "_Default_", "_ASIO_", "_WASAPI (Shared)_", and "_WASAPI (Exclusive)_". A note should be added to WASAPI (Exclusive) and ASIO since they take exclusive access of an audio device, so only osu! can be playing sound from that device.

Linux does not use ASIO, so this will not affect Linux. WASAPI is Windows-exclusive, but does not achieve lower latencies compared to ASIO.

This will also add some dynamic-link dependencies and the use of add-ons from ManagedBass. BASSMIX, BASSASIO, and BASSWASAPI will be used. 

I can work on the code for ASIO & WASAPI support, but some minor design changes may be required.

**Comment by Tom94** [comment]
The latency you describe is not inherent to windows audio mixing. I've had many sound cards, and for most of them there was no perceptible delay, even through DirectSound and an active audio mixer. Therefore, the fault can't lie with windows' audio pipeline. Seems to me like the fault lies with the hardware / driver and not DirectSound.

Of course, if you have a sound card where ASIO / WASAPI actually makes a difference (on all the ones I had, if there was a delay, then it persisted with ASIO / WASAPI; believe me, I tried hard to get rid of it, as I also find it annoying), then feel free to implement it. Design changes can be discussed as they come up.

**Comment by posesix** [comment]
I felt my conclusion is fact because I've worked on and compared low latency code myself and there was significant improvement when using ASIO or WASAPI. I conducted my tests by recording with a microphone and a click input and measured the space between both sounds. Sure enough there was 2ms latency using ASIO and a large 70ms when using DirectSound. The delay is easily noticeable if you make taps after hearing the instrument and getting that feedback.

**Comment by Tom94** [comment]
I'm not claiming your setup doesn't exhibit delay. As I said, feel free to implement ASIO / WASAPI. I'm just saying, that it's most likely a specific problem with your setup (I had similar issues with my old Xonar Essence STX and a bunch of onboard sound cards), and not that of the windows audio mixer. I've got multiple working setups that prove that the latency with DirectSound and the audio mixer turned on can be (and should be) close to none.

**Comment by thomastanck** [comment]
Hi, I'm a chronic sufferer from audio latency. While it was possible for me to fix my setup, it was *EXTREMELY* difficult to do so, having to try out many different drivers and rebooting many times, and sometimes the driver only works under specific circumstances (if I unplug my external monitor and use the laptop monitor, it suddenly stops working!) and sometimes the driver will decide to uninstall itself (I don't have automatic updates enabled, I don't know what's uninstalling it...)

If ASIO / WASAPI is going to remove all this pain, it doesn't actually matter if it's *possible* to fix my setup or not, because it would *actually* fix my setup, and save a decent number of players some grief.

@posesix It'd be great if you could submit a PR for this, else I don't think much can be done ^^;

**Comment by kabaru** [comment]
+1 To this. I've been through 6 months of reinstalls, multiple sound cards, custom drivers, and more. My headphones are connected directly to the sound card so there's definitely no hardware problem. Basically, I've dragged myself through hell to play without latency and am still stuck where I was before. Prior to this I had some drivers which I somehow broke the APO on which had no latency. Obviously I haven't been able to replicate this. My hardware supports ASIO. Please please please add support.

**Comment by peppy** [comment]
In many cases ASIO is not implemented perfectly and can introduce issues (I've experienced this personally in the past). It should not be considered a solves-all solution but instead a fallback for users with problems.

Also please note we do not need +1 comments. We have no issue with ASIO support being added (although do note it is an extra BASS product with a separate license requirement) so feel free to PR support for this.

**Comment by TemTemmie** [comment]
❗️ ppy/osu <- maybe create the issue here

**Comment by Nouko61** [comment]
Audio system is switched to bassmix now, can we raise the priority of this issue?

**Comment by peppy** [comment]
You are welcome to give implementing this a shot at any point, if that's what you're asking!

**Comment by jackiechan420** [comment]
Has there been any update on this?

**Comment by peppy** [comment]
Not yet, but it's on the radar still. Thanks for your patience 😄

**Comment by jackiechan420** [comment]
> Not yet, but it's on the radar still. Thanks for your patience 😄

That's great to hear, will keep patiently waiting for that WASAPI support. Love your work peppy!

**Comment by acid-chicken** [comment]
I recently set up a temporary Linux desktop environment on a Windows machine for recovery purposes. While playing lazer at that time, I was impressed by the reduction in perceived latency. That was enough to convince me that it is worthwhile to support ASIO on Windows.

Given these backgrounds, I'm interested in working on the implementation, so I would appreciate it if you could assign me this issue. If possible, I would appreciate it if you could publish the `ppy.ManagedBass.Asio` [that is currently missing](https://www.nuget.org/packages?q=ppy.ManagedBass).

**Comment by peppy** [comment]
Please start by investigating WASAPI exclusive mode, or reviewing the PR chain which ended on rolling back the WASAPI changes due to some reported latency differences.

ASIO is not a well supported technology on windows. Using WASAPI in a lower latency mode is preferrable.

**Comment by peppy** [comment]
Also we generally don't assign tasks to non-team members. You're welcome to experience and come back with what you come up with though 👍

**Comment by acid-chicken** [comment]
Thank you for your reply.
As you suggested, I've started focusing my work on WASAPI exclusive mode at #6628.

At this point, the patch may be rough around the edges, but it "works," so it's good enough to try things out.

- - -

If anyone is interested in this issue, you can run it on your local Windows machine by pulling the following branch:

<dl><dt>osu-framework</dt><dd><a href="https://github.com/acid-chicken/osu-framework/tree/feat/configurable-audio-backend">feat/configurable-audio-backend</a> in <code>https://github.com/acid-chicken/osu-framework.git</code> (#6628; you may have to rebase with upstream master)</dd><dt>osu</dt><dd><a href="https://github.com/acid-chicken/lazer/tree/feat/configurable-audio-backend">feat/configurable-audio-backend</a> in <code>https://github.com/acid-chicken/lazer.git</code></dd></dl>

**Comment by acid-chicken** [comment]
Fortunately, WASAPI seems to be actually improving latency in my environment.

https://gist.github.com/acid-chicken/9175b94de96a4f64e5b2507d3dac49a7

**Comment by peppy** [comment]
> Fortunately, WASAPI seems to be actually improving latency in my environment.
> 
> https://gist.github.com/acid-chicken/9175b94de96a4f64e5b2507d3dac49a7

FWIW I think your method of testing is unfortunately flawed as mine was when originally releasing WASAPI shared for testing. Basically, the latency *seems* better in shared mode if you test scripted hitsounds vs music track, but it turns out this is just due to better current timestamp tracking which accounts for latency better.

For proper testing you would need to involve triggering hitsounds in response to the audible track, which makes testing slightly harder. The simplest way I've found is to not involve music at all, and just measure time between a loud keyboard key press and the response from speakers / headphones / system output. This could be further optimised by using an input device emulator to trigger the events.

**Comment by acid-chicken** [comment]
Thank you for your feedback.

> FWIW I think your method of testing is unfortunately flawed as mine was when originally releasing WASAPI shared for testing. Basically, the latency seems better in shared mode if you test scripted hitsounds vs music track, but it turns out this is just due to better current timestamp tracking which accounts for latency better.

My initial idea was to measure the time from when an event to play a hitsound (input event in actual play) occurs to when the hitsound plays, using the music track as a clock reference. As you mentioned, this implicitly assumes that the timestamps of the music track are being handled accurately, which may differ from actual play experience latency.

> For proper testing you would need to involve triggering hitsounds in response to the audible track, which makes testing slightly harder. The simplest way I've found is to not involve music at all, and just measure time between a loud keyboard key press and the response from speakers / headphones / system output. This could be further optimised by using an input device emulator to trigger the events.

I tried this idea right away: https://gist.github.com/acid-chicken/5216031bae3bdddb2cb14dc861903aab

**Comment by peppy** [comment]
The testing looks closer to my expectations 👍. You may be able to drop the latency lower with exclusive but buffer underruns become an issue.

Did you happen to test shared WASAPI mode as well? Or was it measured as no difference from Bass with default setup?

**Comment by acid-chicken** [comment]
> The testing looks closer to my expectations 👍. You may be able to drop the latency lower with exclusive but buffer underruns become an issue.

That's good to hear. IMO buffer underruns are an unavoidable trade-off. Since the appropriate buffer size may vary depending on the user's environment and preferences, it might be a good idea to allow users to choose from 2 or 3 different buffer sizes (minimum[, default], safe).

> Did you happen to test shared WASAPI mode as well? Or was it measured as no difference from Bass with default setup?

Yes, unfortunately, this was no different from BASS with the default setup (it was exactly as you had feared regarding my initial test results). I have also added this measurement result to the Gist I mentioned earlier.

**Comment by peppy** [comment]
> Yes, unfortunately, this was no different from BASS with the default setup (it was exactly as you had feared regarding my initial test results). I have also added this measurement result to the Gist I mentioned earlier.

The weird / interesting part is that everything I read in documentation hints at shared mode with correct settings should be able to come very close to exclusive in terms of latency.

If you have any spare bandwidth I would appreciate giving that some time and seeing if you can come up with any reason that's not the case. For instance, you can still reduce the buffer length in custom initialised shared mode, but it doesn't seem to give the expected results, but it might be something we are doing wrong.

Exclusive is great and all, but it doesn't play well with other apps, and I'd really want to avoid having the majority of the player base using it because it gives miles better latency at the cost of having no other system sounds play. If we get a sizeable group of the community using exclusive mode, there would *have* to be an automatic switch to shared mode when the app is not active or not in gameplay or something, which is going to add complexity I'd rather avoid if we can.

**Comment by acid-chicken** [comment]
> The weird / interesting part is that everything I read in documentation hints at shared mode with correct settings should be able to come very close to exclusive in terms of latency.

> If you have any spare bandwidth I would appreciate giving that some time and seeing if you can come up with any reason that's not the case. For instance, you can still reduce the buffer length in custom initialised shared mode, but it doesn't seem to give the expected results, but it might be something we are doing wrong.

I'm willing to take some time to look into this issue within my spare bandwidth. First, I tried changing the category designation from raw mode to another one, stopping event-driven, or explicitly passing buffer size and period size as device minimum values, and verified with the previous method, but none of them effect.

Also, I noticed that by using `BASS_WASAPI_GetInfo`, we can check the actual buffer size being used after initialization. This value returns 1,152 in exclusive mode and 7,872 in shared mode (regardless of flags). According to the documentation, this refers to the byte size of the buffer length. Since the format type is 24-bit integer for exclusive mode and 32-bit floating point for shared mode, converting by sample format and channel count gives 192 samples (\~= 4.4 ms) for exclusive mode and 1,968 samples (\~= 22.3 ms) for shared mode.

Comparing with the previous measured values, it seems that shared mode has a penalty of about 12 ms in addition to the buffer size, compared to exclusive mode.

That's all I have worked on and found out today. I'll try again as soon as I come up with any ideas.

> Exclusive is great and all, but it doesn't play well with other apps, and I'd really want to avoid having the majority of the player base using it because it gives miles better latency at the cost of having no other system sounds play. If we get a sizeable group of the community using exclusive mode, there would have to be an automatic switch to shared mode when the app is not active or not in gameplay or something, which is going to add complexity I'd rather avoid if we can.

I agree that we should avoid having exclusive mode become the majority. That said, since we are close to a dead end at this point, I think it is worth checking how the buffer sizes are set in each mode in other people's environments.

**Comment by acid-chicken** [comment]
> Also, I noticed that by using BASS_WASAPI_GetInfo, we can check the actual buffer size being used after initialization. This value returns 1,152 in exclusive mode and 7,872 in shared mode (regardless of flags). According to the documentation, this refers to the byte size of the buffer length. Since the format type is 24-bit integer for exclusive mode and 32-bit floating point for shared mode, converting by sample format and channel count gives 192 samples (\~= 4.4 ms) for exclusive mode and 1,968 samples (\~= 22.3 ms) for shared mode.

First of all, to correct yesterday's statement, I intentionally increased the buffer size to check the correlation between the number of samples and the buffer size used at initialization. It seems that regardless of the format used at initialization, exclusive mode is also calculated as 32-bit. Therefore, the correct value is 144 samples (\~= 3.3 ms) for exclusive mode, not 192 samples (\~= 4.4 ms). There was no calculation error in shared mode *sadly*.

Next, I tried some other devices that I could quickly access. Using several audio interfaces on the same machine showed slightly better values by about 1-2%. This is probably due to the ability to determine the buffer size in finer units. I also tried two other machines, both of which showed a worse value of 8,448.

- - -

BTW exploring the documentation on low-latency audio, I found a page titled "Low Latency Audio" on Microsoft Learn. In the [Audio stack improvements](https://learn.microsoft.com/en-us/windows-hardware/drivers/audio/low-latency-audio#audio-stack-improvements) section of this page, it is stated as follows:

> Windows 10 and later have been enhanced in three areas to reduce latency:
>
> 1. All applications that use audio will see a 4.5-16 ms reduction in round-trip latency (as was explained in the section above) without any code changes or driver updates, compared to Windows 8.1.
>     1. Applications that use floating point data will have 16-ms lower latency.
>     1. Applications that use integer data will have 4.5-ms lower latency.
> 1. Systems with updated drivers will provide even lower round-trip latency:
>     1. Drivers can use the low latency DDIs to report the supported sizes of the buffer that is used to transfer data between Windows and the hardware. Data transfers don't have to always use 10-ms buffers, as they did in previous Windows versions. Instead, the driver can specify if it can use small buffers, for example, 5 ms, 3 ms, 1 ms, etc.
>     1. Applications that require low latency can use the low latency audio APIs (AudioGraph or WASAPI), to query the buffer sizes that are supported by the driver and select the one that will be used for the data transfer to/from the hardware.
> 1. When an application uses buffer sizes below a certain threshold to render and capture audio, Windows enters a special mode, where it manages its resources in a way that avoids interference between the audio streaming and other subsystems. This will reduce the interruptions in the execution of the audio subsystem and minimize the probability of audio glitches. When the application stops streaming, Windows returns to its normal execution mode. The audio subsystem consists of the following resources:
>     1. The audio engine thread that is processing low latency audio.
>     1. All the threads and interrupts that have been registered by the driver (using the low latency DDIs that are described in the section about driver resource registration).
>     1. Some or all of the audio threads from the applications that request small buffers, and from all applications that share the same audio device graph (for example, same signal processing mode) with any application that requested small buffers:
>         1. AudioGraph callbacks on the streaming path.
>         1. If the application uses WASAPI, then only the work items that were submitted to the Real-Time Work Queue API or MFCreateMFByteStreamOnStreamEx and were tagged as "Audio" or "ProAudio".

And in the [FAQ](https://learn.microsoft.com/en-us/windows-hardware/drivers/audio/low-latency-audio#faq) section, under "If a driver supports small buffer sizes, will all applications in Windows 10 and later automatically use small buffers to render and capture audio?", it is stated as follows:

> No, by default all applications in Windows 10 and later will use 10-ms buffers to render and capture audio. If an application needs to use small buffers, then it needs to use the new AudioGraph settings or the WASAPI IAudioClient3 interface, in order to do so. However, if one application requests the usage of small buffers, then the audio engine will start transferring audio using that particular buffer size. In that case, all applications that use the same endpoint and mode will automatically switch to that small buffer size. When the low latency application exits, the audio engine will switch to 10-ms buffers again.

In other words, my understanding at this point was, if a driver supports it, when someone in the process requests a small buffer size in WASAPI shared mode while meeting the above requirements, other processes sharing the same endpoint will also use that buffer size.
I found a project called https://github.com/spddl/LowAudioLatency that does exactly this, so I tried it. This application claims that the available buffer size has both an upper and lower limit of 10 ms, and it actually initializes with 10 ms. Then, even when I launched osu! in this state, there was no change in the BASSWASAPI buffer size.

In short, it seems that none of the machines or devices I have available support low-latency audio in WASAPI shared mode.
And, based on what I've learned so far, I speculate that the new initialization method using `IAudioClient3` initializes with a buffer size synchronized with the master's buffer size in shared mode. BASSWASAPI, on the other hand, likely uses an older initialization method (such as `IAudioClient2`?), which probably manages a separate buffer from the master and can only be initialized at a minimum of 22 ms at this point. It seems that the master's buffer size is added as a penalty on top of this (the approximately 12 ms penalty observed in the previous verification is probably due to this).

To reiterate, this is merely an inference based on the circumstances, so please do not rely on it.

That's all for today. I'll keep trying things out as ideas come to me.

**Comment by peppy** [comment]
> I found a project called https://github.com/spddl/LowAudioLatency that does exactly this, so I tried it. This application claims that the available buffer size has both an upper and lower limit of 10 ms, and it actually initializes with 10 ms. Then, even when I launched osu! in this state, there was no change in the BASSWASAPI buffer size.

Does this imply you successfully halved the buffer size in shared mode using this method?

**Comment by acid-chicken** [comment]
> > I found a project called https://github.com/spddl/LowAudioLatency that does exactly this, so I tried it. This application claims that the available buffer size has both an upper and lower limit of 10 ms, and it actually initializes with 10 ms. Then, even when I launched osu! in this state, there was no change in the BASSWASAPI buffer size.
>
> Does this imply you successfully halved the buffer size in shared mode using this method?

No, this does not bring any improvement at all. From here, I will state my speculation on this point.

The shared mode used by BASSWASAPI, that is, the traditional WASAPI shared mode, probably has a structure like the one described below:

```
Device ◄─── Master (10 ms)
(10 ms)       ▲
              ├─── App1 (100 ms)
              ├─── App2 (50 ms)
              ├─── App3 (30 ms)
              ... Each app has its own buffer size. Double buffering enables support for flexible buffer sizes?
```

And the low-latency audio mode, which is said to have been added from Windows 10, probably has a structure like this:

```
Device ◄─┬─ Traditional Master (10 ms)
(period  │    ▲
 will be │    ├─── App 1 (100 ms)
 at the  │    ├─── App 2 (50 ms)
 minimum │    ├─── App 3 (30 ms)
 it sup- │    ... Each app has its own buffer size. Double buffering enables support for flexible buffer sizes?
 ports;  │
 e.g.    ├─ Low Latency App 1 (10 ms)
 1 ms)   ├─ Low Latency App 2 (5 ms)
         ├─ Low Latency App 3 (3 ms)
         ... Smaller buffer sizes that are units of the device's minimum period may be available?
```

As I mentioned earlier, I couldn't try low-latency audio in my environment, so this is largely speculative, but based on the information that low-latency mode achieves latency equivalent to or better than exclusive mode[^1], I think it has this structure or something close to it.

[^1]: https://en.namu.wiki/w/WASAPI#s-3.3

**Comment by hwsmm** [comment]
To add some here, with my backend (#6002) and https://github.com/libsdl-org/SDL/pull/9683 (IAudioClient3), I could get 15ms latency on the game. The methodology was mostly the same with what peppy mentioned, so it's probably fine to think that shared mode can be a lot better if it is correctly configured.

The problem is that most proprietary sound card drivers don't support this. One way to get around this is to use Microsoft HD Audio driver instead, but it is not so accessible and doesn't guarantee low latency from what I've seen. There really isn't an easy way to achieve low latency audio on Windows.

**Comment by acid-chicken** [comment]
> BASSWASAPI, on the other hand, likely uses an older initialization method (such as IAudioClient2?), which probably manages a separate buffer from the master and can only be initialized at a minimum of 22 ms at this point.

To correct this, according to a [forum post](https://www.un4seen.com/forum/?msg=133785), BASSWASAPI uses `IAudioClient3` when the buffer size is set to 0 during initialization. In other words, the 22 ms buffer size is not due to differences in initialization methods but seems to be caused by some other unknown factor.

> The problem is that most proprietary sound card drivers don't support this. One way to get around this is to use Microsoft HD Audio driver instead, but it is not so accessible and doesn't guarantee low latency from what I've seen. There really isn't an easy way to achieve low latency audio on Windows.

Unfortunately, I think this is indeed the case. It seems that [MOTU is one of the few vendors known to have implemented WaveRT in their drivers](https://blog.nirbheek.in/2018/03/low-latency-audio-on-windows-with.html), but in any case, we cannot expect low-latency mode to be used in userland. Therefore, our goal is to achieve a 10 ms buffer size.

**Comment by acid-chicken** [comment]
> > BASSWASAPI, on the other hand, likely uses an older initialization method (such as IAudioClient2?), which probably manages a separate buffer from the master and can only be initialized at a minimum of 22 ms at this point.
>
> To correct this, according to a forum post, BASSWASAPI uses IAudioClient3 when the buffer size is set to 0 during initialization. In other words, the 22 ms buffer size is not due to differences in initialization methods but seems to be caused by some other unknown factor.

In order to further investigate this phenomenon, I captured the WASAPI API calls and found that BASSWASAPI first calls `IAudioClient3::GetSharedModeEnginePeriod` to obtain the range of supported buffer sizes.

This method returns a value in `pMinPeriodInFrames` that is smaller than `pDefaultPeriodInFrames` in environments that support Low Latency Mode, but all values are the same in environments that do not.
If a smaller buffer size is available, it then calls `IAudioClient3::GetCurrentSharedModeEnginePeriod` to get the current buffer size, and likely subsequently calls `IAudioClient3::InitializeSharedAudioStream`.
However, in environments where the range of buffer sizes obtained from `IAudioClient3::GetSharedModeEnginePeriod` are all the same, meaning that the environment does not support Low Latency Mode, it aborts the initialization using `IAudioClient3` and falls back to calling `IAudioClient::Initialize`.

From this, it seems that there is currently no way to make BASSWASAPI use `IAudioClient3` in environments that do not support Low Latency Mode.

**Comment by acid-chicken** [comment]
> Exclusive is great and all, but it doesn't play well with other apps, and I'd really want to avoid having the majority of the player base using it because it gives miles better latency at the cost of having no other system sounds play. If we get a sizeable group of the community using exclusive mode, there would have to be an automatic switch to shared mode when the app is not active or not in gameplay or something, which is going to add complexity I'd rather avoid if we can.

After trying to implement the switching between exclusive and shared modes, I found that it was possible to do so while avoiding a significant increase in complexity, at least compared to my initial expectations. The implementation can be found here (most of the changes are just namespace refactoring. the key point of the diff is [osu.Game/OsuGame.cs](https://github.com/acid-chicken/lazer/blob/625f7bd064a26102ff154627d8bce3d91c56887d/osu.Game/OsuGame.cs#L455-L506)): https://github.com/acid-chicken/lazer/commit/625f7bd064a26102ff154627d8bce3d91c56887d#diff-df88cc06584b79009417332730decf4c99f38f5fd5f9af1fc9f5dad547cc9814R455-R506

Does this alleviate your concerns regarding the current direction?

**Comment by peppy** [comment]
I would see it being a framework side implementation. I cannot imagine any framework application *not* wanting this automatically done when window becomes inactive. Does it result in an audible skip? What's with all the `using` updates? Let's avoid that too.

**Comment by acid-chicken** [comment]
> I would see it being a framework side implementation. I cannot imagine any framework application not wanting this automatically done when window becomes inactive. Does it result in an audible skip? What's with all the using updates? Let's avoid that too.

Audible skips occur slightly and cannot be eliminated IMO. In my measurements, it is about the length of adding 20 ms to the buffer size.

The rest has been fixed in #6628.

**Comment by LastExceed** [comment]
I have spent an unreasonable amount of time researching this topic, and might have some exciting results to share:

<img width="1534" height="770" alt="Image" src="https://github.com/user-attachments/assets/c150142d-8eef-4379-b6d1-541eda6d4afd" />

## What is being measured exactly?
* mouse button (physical click) -> headphone output (hitsound)
* in [this](https://github.com/LastExceed/wasapinimal/blob/main/src/main.rs) minimal test app, which i created specifically for this purpose (and in osu for comparison)
* using winapi directly (no libraries beyond stdlib and ffi bindings)

## Why not measure RTL?
round-trip latency via hardware loopback could indeed eliminate mouse input latency from the equation. however:
* we'd just replace it with audio input latency instead, which is likely worse
* click-to-hitsound is most representative of what players ultimately care about
* start and end can be observed in the same signal, guaranteeing correct temporal correlation
* I am confident that the involved non-audio latencies of my system are negligible (i can elaborate on this if desired)

## Why is there so much variance?
* My app only registers user input once per period (to keep the code simple), which introduces that much variance, due to cycle alignment
  * WASAPI shared: 10 ms
  * WASAPI exclusive: 3 ms
  * ASIO: 1/6 ms
* I can't speak for the game, its variance surprises me, considering that it tracks input timings religiously

## Why does compatibility mode outperform default?
* I don't know
* The hover text claims the opposite
* I am anxious about potential methodology errors, or quirks in my system, because of this
* I wish I at least knew what that checkbox actually does on a technical level ("legacy audio backend" is not very specific)
* Hopefully a team member with code access can provide some insights

## What does "with delay" mean?

Apart from cycle alignment, another drawback of event driven mode is that processing always happens at the very start of a period, resulting in 1 full period of additional latency. To demonstrate the potential of polling (which doesn't suffer from this problem) without actually having to implement it, I have added a `sleep()` right after the event is signaled, in order to shift all processing (notably including input registration) towards the end of the period. This is obviously a massive hack due to risk of oversleeping, and it doesn't even provide the best possible latency yet, but it clearly shows that there is plenty room for improvement still, even after having cut the latency in half already

## `IAudioClient` vs `IAudioClient2` vs `IAudioClient3`
I tried switching between them, while keeping everything else the same. There seems to be no difference. But the way I understand COM interfaces, we shouldn't expect one anyway, as they most likely all point to the same implementation. They're really just that: Interfaces.

*(Fun fact, this is literally where the C# keyword and the associated naming convention come from. in C/C++ `interface` isnt actually a language keyword, but a winapi macro that resolves to either `class` or `struct` depending on C vs C++)*

The newer ones do expose some functions that seem to be relevant for latency optimization though (notably `IAudioClient2::SetProperties()` and `IAudioClient3::InitializeSharedAudioStream()`), but I have yet to look into that. Will report back when I find something.

## Conclusion

I was very surprised to realize how simple the code turned out. However, getting there was anything but.

Winapi has a tendency to feel very arcane, and you end up spending many hours sifting through obscure places like [this monster of a source file](https://github.com/PortAudio/portaudio/blob/master/src/hostapi/wasapi/pa_win_wasapi.c) which was clearly written by someone who is intimately familiar with windows' internals, desperately trying to figure out how PortAudio manages to maintain ~20ms in shared mode. And now that I finally managed to pull it off myself, I can't even explain what took me so long (as is tradition with winapi it seems).

> The weird / interesting part is that everything I read in documentation hints at shared mode with correct settings should be able to come very close to exclusive in terms of latency.  
> If you have any spare bandwidth I would appreciate giving that some time and seeing if you can come up with any reason that's not the case. For instance, you can still reduce the buffer length in custom initialised shared mode, but it doesn't seem to give the expected results, but it might be something we are doing wrong.

I have a working theory for this, but I need to do some more testing (with BASS in particular) before I'm comfortable pointing fingers

---

@acid-chicken

> In order to further investigate this phenomenon, I captured the WASAPI API calls and found that BASSWASAPI first calls `IAudioClient3::GetSharedModeEnginePeriod` to obtain the range of supported buffer sizes.

 Can you tell me how you did this? It sounds very useful!

Also - I know its off-topic, but I can't not mention this - are you my doppelgänger?? I
* kept an eye on this issue ever since it was created
* saw no progress in 8 years
* finally decided to give it a shot myself about a month ago

and suddenly there is another person who
* shares the same oddly specific interest in audio latency as me
* tries to implement the same long standing issue as me
* started doing so on almost the **exact same day** (-2) as me

and above all else: when using dark mode EVEN THE PROFILE PICTURE LOOKS THE SAME!

spooky @.@

**Comment by LastExceed** [comment]
This research has turned into an obsession. In the past week alone I've easily spent another ~60h drilling into WASAPI and BASS, and the results are nothing to sneeze at:

![](https://i.imgur.com/2GBuPQG.png)

Don't believe it? I sure didn't, so [here](https://github.com/user-attachments/files/22856115/raw.wav) is a recording (music played from youtube to ensure shared mode), and [here](https://github.com/user-attachments/files/22856118/trimmed%2Bslowed.wav) is that same recording trimmed and slowed 0.08x (so you can tell apart clicks and hitsounds)

Most importantly: Unlike the data in my last post, for which I used a dedicated pro audio interface (since it natively supports ASIO), these new results here were achieved entirely on the integrated (Realtek®) audio jack of my (very basic) motherboard, and should thus be reproducible by a wide audience.

There is a lot to go through, so lets get started!

# BASS & BASSWASAPI

When configuring the buffer size in IAudioClient, the specified size is only a minimum, not an exact value. And when requesting a very small buffer, WASAPI tends to provide a bigger one. For instance, when requesting a 10ms buffer (=the default minimum periodicity in shared mode), WASAPI actually allocates a 22ms buffer. This is works because, unlike ASIO or BASS' intermediate buffers, which use an alternating buffer system, WASAPI uses something more akin to a queue. However when you're used to alternating buffer systems, a very natural thing to do is to just always write however much data there is space available in the buffer for. And since this will still result in the expected behaviour of writing 10ms of data every 10ms, it is very easy to miss the fact that you're actually queueing up an additional 12ms (=22-10) worth of data. Does that number ring a bell? If not, Ctrl+F "12 ms" and scroll up.

BASS and BASSWASAPI do exactly that. Whether by accident or on purpose I'm not sure, but the point is that we have no control over this. I've tried cheesing the system by configuring a callback that never provides more than 10ms worth of data, regardless of how much BASS asks for, but unfortunately BASS already fills the WASAPI buffer with silence during initialization, so all this achieves is preventing catchup after falling behind, resulting in more underruns. (Not to mention the fact that we'd once again rely on an unstable hack, which I would very much like to avoid)

There's more. At some point throughout this journey I have observed that BASS has 8ms more latency than BASSWASAPI, and I had to dig really deep to figure out why. I eventually noticed that changing [`BASS_CONFIG_DEV_BUFFER`](https://www.un4seen.com/doc/#bass/BASS_CONFIG_DEV_BUFFER.html) appears to have no effect on latency (in either direction), and after finally figuring out a way to capture BASS' calls to the WASAPI API (thank you @acid-chicken for the idea, it has been very enlightening indeed!), I've discovered that BASS always requests the default 30ms buffer size no matter what I have configured. 30ms vs 22ms perfectly explains the 8ms difference between BASS and BASSWASAPI I had observed. [The documentation](https://www.un4seen.com/doc/#bass/BASS_CONFIG_DEV_BUFFER.html) states:
> The buffer length needs to be a multiple of, and at least double, the device's update period, which can be set via the [`BASS_CONFIG_DEV_PERIOD`](https://www.un4seen.com/doc/#bass/BASS_CONFIG_DEV_PERIOD.html) option. The buffer length will be rounded up automatically if necessary to achieve that.

Perhaps it is rounding up the 22ms to 30ms, but then again I can't increase the value either, so maybe this is a bug. Luckily we don't have to worry about this, as we can just use BASSWASAPI which doesn't use this value in the first place.

# Native WASAPI

To get rid of the remaining 12ms without having to replace BASS wholesale we could grab the output of the global mixer stream and shove into WASAPI ourselves. I have implemented a proof-of-concept of this in #6649.

But what else could we do with direct WASAPI access? In my last post I mentioned a couple loose ends I had yet to investigate, which I have now done:

## IAudioClient2

This one doesn't bring too much to the table, mostly some hardware offloading stuff for low-power devices which is explicitly said to incur a lot of latency, and thus of no interest here. It does introduce the [`IAudioClient2::SetProperties()`](https://learn.microsoft.com/en-us/windows/win32/api/audioclient/nf-audioclient-iaudioclient2-setclientproperties) method, where we can enable "raw mode" which lets us skip [audio processing objects](https://learn.microsoft.com/en-us/windows-hardware/drivers/audio/audio-processing-object-architecture) (shoutout to [EqualizerAPO](https://sourceforge.net/projects/equalizerapo/) o7), but those usually don't cause much latency to begin with, and I'm not even sure we *want* to skip those at all. Maybe as a user setting, but thats a topic for another day.

## IAudioClient3

As mentioned earlier in this thread, the star of the show here is [`IAudioClient3::InitializeSharedAudioStream()`](https://learn.microsoft.com/en-us/windows/win32/api/audioclient/nf-audioclient-iaudioclient3-initializesharedaudiostream) (the other methods are basically supplements for this one). At some point I had already more or less written it off as "just a convenience wrapper around [`IAudioClient::Initialize()`](https://learn.microsoft.com/en-us/windows/win32/api/audioclient/nf-audioclient-iaudioclient-initialize)", because in my testing they performed the exact same, but then I re-read this thread and realized that my drivers just didn't support its benefits.

# The critical detail

@hwsmmm had it spot-on when they said
> The problem is that most proprietary sound card drivers don't support this. One way to get around this is to use Microsoft HD Audio driver

because from what I have gathered it really does seem to be supported by OEMs quite rarely. I tried the aforementioned Microsoft driver (installation instructions can be found [here](https://learn.microsoft.com/en-us/windows-hardware/drivers/audio/low-latency-audio#measurement-tools), buried faaaaar down the page in the "measurement tools" section for some reason) and MY GOD does it make a difference. Suddenly shared mode allows periodicities down to 2.5ms, yielding the latencies shown in at the beginning, and exclusive mode rivals the performance of native ASIO. Further more, it seems to provide a very granular audio clock, with individual sample precision even at 192khz. I don't know much about drivers, but I'm honestly wondering why Realtek even bothers to make their own, instead of defaulting to this one.

I immediately went back to test osu!stable once more, and while the driver change alone only improved latency by ~2ms there (both with and without compatibility mode), it enables the [LowAudioLatency](https://github.com/spddl/LowAudioLatency/) tool @acid-chicken brought up earlier to acutally makes a difference - a quite substantial one: Another -15ms! My click-to-speaker latency in osu!stable is now 17-27ms (down from 34-44ms) just by switching drivers and running a tiny commandline tool in the background.

# Back to BASS

Did I already mention that I even managed to find the source of the ±5ms variance in sample latency during this odyssey? Well now I did. Its another quirk in BASS. Details in #6650 because its a bit out of scope, but I'm bringing it up because the following happens to also mitigate that problem somewhat as a side effect:

## How do we actually make use of this?

Amusingly, the answer has already been posted in this thread (ref @acid-chicken for the 3rd time), I guess we just didn't fully understand it yet:
> according to a [forum post](https://www.un4seen.com/forum/?msg=133785), BASSWASAPI uses `IAudioClient3` when the buffer size is set to 0 during initialization

Yes, really. That's all it takes. And don't worry, this isn't just an obscure implementation detail. The behaviour is by design, as documented [here](https://www.un4seen.com/doc/#basswasapi/BASS_WASAPI_Init.html).

## Doesn't a period as tiny as 2.5ms lead to constant underruns?

I don't think so actually. I believe the underruns in the current BASSWASAPI implementation (disabled as of #6093) are mainly a result of cycle misalignment between BASS' update rate and WASAPI's data consumption, which can be fixed by simply enabling `BASS_WASAPI_EVENT` in [BASS_WASAPI_Init()](https://www.un4seen.com/doc/#basswasapi/BASS_WASAPI_Init.html). Frankly, I find BASSWASAPI's design a bit questionable here. It defaults to polling, but hides the alternatives to using a callback in the form of surrogate callback constants (`WASAPIPROC_PUSH` and `WASAPIPROC_BASS`)

*fun fact: Using negative function pointers as surrogate values is such a hack that the Rust compiler outright rejects them even in unsafe code (due to pointer provenance technicalities)*

This is probably what encouraged the janky polling + callback configuration we currently have (it performs poorly on my machine too). I've switched to event driven mode with 2.5ms period for a ~2h gameplay session and don't remember any glitches except during startup, but of course this might just be my machine. Looking forward to see how well others fare.

## What happened to the 12ms from premature buffer filling?

The issue persists, but on a much smaller scale. With a periodicity of 2.5ms, WASAPI allocates a 5.5ms buffer, which is just 3ms more than we need. The final click-to-output latency I get on my PR is 11.5-15.5ms. If we really want to push it, we could save those 3ms with the native WASAPI trick I mentioned earlier. Dunno where the last 2ms (compared to my minimal test app) come from, but neither do I think it matters. At this point we might as well look into exclusive mode again.

Speaking of:

## Does this obsolete exclusive mode?

No. Without good drivers we're stuck at 20+ ms in shared mode, so exclusive mode still provides relevant gains there. But we probably don't need to worry about it becoming a meta anymore. At best I could see a "Microsoft HD Audio Driver" meta emerging, but I think that'd be much less of a problem.

Further more, exclusive mode is basically a free upgrade for people with multiple sets of speakers, mixing capable hardware, or just the patience and terchnical expertise to configure some virtual loopback shenanigans, so I think it would really be a shame if the leading rythmgame at the time (no, GD is not a rythm game, fight me) didn't support it.

## What about ASIO?

I'd say there is still a case to be made for ASIO even, as its performance remains unmatched on most hardware that supports it.

There are also a few software ASIO drivers (e.g. [Asio4all](https://asio4all.org/), [FlexAsio](https://github.com/dechamps/FlexASIO/), [Asio2ks](http://www.asio2ks.de/)) built directly on top of WDM-KS, which might be able to squeeze out a bit more latency than WASAPI on suboptimal drivers, but I haven't looked into this much.

## WDM-KS ?

Stands for "Windows driver model kernel streaming". This is basically the layer below WASAPI. I hesitate to even mention the idea of implementing it as an audio backend directly because of how wrong that sounds, but then again that's just my gut feeling. I don't actually know much about it, nor what the implications would be.

---

Since we'll probably be adding the infrastructure for supporting multiple audio backends anyway (#2784, #6002, #6628), implementing 1 or 2 more shouldn't be too difficult. But the priority goes down for sure.

# TL;DR

* Can't do much better than our current ~40ms with BASS alone
* Can't get below ~20ms without better drivers
* Good OEM drivers are rare
* Windows ships with a good universal driver, but eh
* Good drivers can bring shared mode below 10ms
* Exclusive mode with good driver rivals native ASIO
* I found the source of sample latency variance
* I think I fixed BASSWASAPI instability

And in the most poetic fashion imaginable, this multi-hundred hour research that has completely consumed me for the past idontevenknowhowlong finally concludes with a PR of complementing size:

[A single line of code.](https://github.com/ppy/osu-framework/pull/6651)

**Comment by peppy** [comment]
Thanks again for your research.

One take-away from this is that we may want to inform users if they aren't using MS drivers for their audio card (and can) with text in the settings UI.

**Comment by LastExceed** [comment]
On the one hand, I'm a bit apprehensive about actively encouraging users to change their drivers until we thoroughly understand the implications. the MS driver just seems too good to not have any downsides. There gotta be a reason for its obscurity

But on the other hand, the improvement is way to attractive to miss out on.

i suppose an indicator for whether or not the current driver is "low latenct ready" (= supports <10ms periodicities) with a link to the wiki where we have enough space for an elaborate disclaimer about the uncertain and experimental nature of switching to the MS one is appropiate. Since its signed by Microsoft, and bundled with the OS by default, there shouldnt be any security concerns, but I could imagine that it breaks stuff like OEM configuration software or propietary stream formats and whatnot, so we need to make sure that users will look out for stuff like this, and know how to revert the change if needed

**Comment by Spok5508** [comment]
I have a lot of experience with trying to force MS drivers on various cards from various vendors, there's definitely cases where the driver breaks a lot of functionality (for example, many of the Soundblaster cards only work through 1 port or spdif with these drivers).

> i suppose an indicator for whether or not the current driver is "low latenct ready" (= supports <10ms periodicities) with a link to the wiki where we have enough space for an elaborate disclaimer about the uncertain and experimental nature of switching to the MS one is appropiate.

I think this is the best way to go about things. I think it would be amazing if the client could tell the user what the smallest periodicities can be on their current hardware/driver setup (possible with `GetSharedModeEnginePeriod`), and then give the user the option to enable this period to get the absolute lowest possible latency in shared mode (maybe with an added disclaimer of potential underruns or audio pops).

**Comment by mirh** [comment]
If the only requirement for low latency mode is WaveRT, then why would realtek require to revert back to the basic HDA driver?
I couldn't check latency, but [AFAIU](https://www.kvraudio.com/forum/viewtopic.php?t=224491#p3192855) even Vista-era ones had `KSCATEGORY_REALTIME`.
(btw it's [unclear](https://community.osr.com/t/how-to-intercept-and-modify-microphone-speaker-data-stream-in-audio-filter-driver-kmdf/58976/4) how much USB devices could support, at least with the default drivers)

As for WDM-KS, you are indeed correct that is quite a can of worms to re-implement from scratch.. But you don't have too.
[Portaudio](https://github.com/PortAudio/portaudio/blob/master/src/hostapi/wdmks/pa_win_wdmks.c) took decades of blood to come up with that fat boilerplate.
Admittedly you could also just plug-in asio, which then could use the various xASIOx wrappers. But you would have to put up with Steinberg's sdk, eventually just to use what are still WDM wrappers that aren't even under your own control (exception being flexasio that does indeed use portaudio beneath, but at that point you might as well cut off the middleman).

**Comment by acid-chicken** [comment]
@LastExceed Hello! I'm glad to greet my doppelgänger in the most spooky season of the year.
Also, albeit belated, thank you for your findings which complemented each other and enhanced the reliability of the results.

- - -

Regarding my environment, it is one of those that gets left behind from low-latency audio mode even after overwriting the motherboard's audio chip driver with the Microsoft HD Audio Driver.
Considering that other sound cards and audio interfaces also have some issues or get left behind, I still think there is value in supporting exclusive mode.

Also, most of the audible skips when using small buffers occur on the Song Select screen, which I suspect is due to the decoding of the first chunk of the audio track not keeping up. I haven't had much time lately, but I plan to investigate this when I have some spare time.

**Comment by peppy** [comment]
>  the MS driver just seems too good to not have any downsides. There gotta be a reason for its obscurity

In my experience, there generally isn't. The third party drivers tend to add support for their extended control panel and don't meet the specs which microsoft ones are ensured to. If you are using hardware from a competent company with competent driver developer, the proprietary is usually better... but as we've seen that's definitely not the case with realtek.

This also applies to things like network adapters, and sometimes (usually more rarely) GPUs.

**Comment by LastExceed** [comment]
since #6651 went live, i had a bunch of people contact me (mainly for glazing lol) and was on stream with btmc to talk about the tech, and so far i havent heard a single case where switching drivers caused any problem other than the (expected) breakage of OEM configuration software, so looks like you are correct. could of course be that people with problems just reverted back and dont want to tell anyone theyre "missing out", but even that would still imply that doing so is easy enough to not be of concern, so im starting to feel more comfortable about openly recommending the driver change. will keep looking out for trouble stories ofc

im currently working on understanding and refactoring the audio part of osu-framework (i may be an expert on the native libs/apis, but i'm still mostly clueless about the architecture of osu-framework itself, can you believe it) so i can add support for additional playback backends (just like we have multiple graphics backends to choose from) including some way to display backend specific info (design TBD), e.g. ASIO drivers allow manufacturers to specify the expected internal latency of the hardware. for WASAPI i would add supported periodicity range, based on which we can then determine whether or not to suggest a driver change to the user

once again thank you @acid-chicken for the work you've already done on this in #6628, which i am using for reference a lot (so rest assured, your work was not in vain). i feel a bit bad about getting all the glory from the community here, because you definitely deserve some credit too. @hwsmm 's #6002 is also still on my radar, but that one includes reworking DSP, which i consider out of scope for now

**Comment by peppy** [comment]
With WASAPI now working as i originally expected it should, we are highly unlikely to add more complexity to the audio backend (especially something like multiple audio backends).

If you have any interest in going further with WASAPI exclusive or ASIO, keep in mind that they have serious usability concerns which will likely block them from being candidates for osu!/osu!framework (discussed elsewhere). As already mentioned on discord, we are not interested in multiple audio backends. If we were to ever change things, it would be using something like acid-chickens work as a complete replacement, not a choice.

**Comment by LastExceed** [comment]
> discussed elsewhere

care to be little bit more specific? without any context this isn't very constructive

**Comment by peppy** [comment]
Discussions have happened in real life, discord, other threads linked by spaceman earlier (and probably more). I'd recommend searching around previous PRs, issues (closed too) and blaming through commits to get a better understanding.

Further work in this direction is not on our roadmap and – as mentioned in discord – even if features are implemented, assume it to be discardable work from the outset so there's no disappointment.

**Comment by bdach** [comment]
For the record, the following is a more charitable depiction of where I personally stand on "multiple audio backends":

- I don't know what is specifically referred as to "audio backend" when it comes to code, even when limited to unspecific "playback" as @LastExceed did on discord. Is every single audio-related class (tracks/samples) going to receive a "backend"-specific implementation? What warrants doing that? What is the backend going to be? What is it going to be doing different?
- Is there a real reason to introduce "backend" abstractions? A real reason as far as I'm concerned would entail having situations where sometimes "backend" A performs worse than B, but sometimes performs better, in some respect, or there are other tradeoffs involved that actually warrant being able to switch things. Which *is* very much the case with graphics. If there's no such reason to do it then "backends" shouldn't exist, and the only implementation in source should be the one that is the best at all times, which is sort of where (as far as I'm aware) the current plugin-based implementation of WASAPI is, and why I'm more open to just deleting BASS in favour of SDL audio if it improves the situation all around (which isn't clear either).
- An additional caveat to the above is that determining whether there is a "real reason" in terms defined above will likely result from pushing the change to users and seeing what happens. The current improved WASAPI hasn't even seen a full proper lazer release. I don't discount the possibility than when pushed to a larger userbase, issues will surface that may warrant reevaluating the audio situation, but I won't put wagers on it *before* said release happens.

If some of the above can be elaborated on / explained to my out-of-touch self, I may be more open to the possibility, and may be more willing to provide commitments from my side that any further work on these "backends" is going to be considered further. But I am *very* aggressive on cutting complexity when it's not warranted otherwise.

**Comment by mirh** [comment]
Yeah, complexity just for its own sake is pointless. 

Though besides perhaps audio devices that don't even expose a HDA endpoint, why you may not want to stick with the basic driver seems easy to guess.
Particularly in this day and age of spatial audio (to be sure you aren't running your DTS:X Ultra or Sound Unbound license without the OEM cruft) and DSP-driven array speakers.
Plus I guess even if it's just an extra 1 or 2 ms (tops) on the table to improve, it's nice to even pull off that ease of mind.

So that's why besides WASAPI only whatever had WDM-KS/ASIO (which would also have the added benefit of bypassing whatever fuck technical limitation vendor drivers have?) seems to make sense.

**Comment by acid-chicken** [comment]
The abstraction introduced in #6628 initially aimed to organize the audio processing structure to avoid worsening issues mentioned in places like https://github.com/ppy/osu-framework/blob/5acb295c0955e3f5f76c7a3b9f17299a8c36ff03/osu.Framework/Threading/AudioThread.cs#L122 when working on it.

To ensure compatibility with BASSASIO and/or SDL, this abstraction needs to be advanced anyway. However, if there is no opportunity for that, of course, it is also possible to maintain the current structure. In that case, I'll isolate WASAPI exclusive mode as a simple option from the PR.

**Comment by bdach** [comment]
Refactors to move logic out of places where it probably shouldn't be, *hopefully* isn't the same thing as 1.8k line diff stat refactors of half the audio subsystem. A more gradual cleanup of the code, without "pluggable modules" or anything, might be viewed more amicably.

**Comment by LastExceed** [comment]
*[link to discord conversation for full context](https://discord.com/channels/188630481301012481/589331078574112768/1432364602720518265)*

---

I'm really disappointed to receive such a prejudiced lecture in response to a trivial technical question. I know you probably mean no evil, but please try to be more mindful about your attitude in the future, because this has been very uncomfortable and discouraging. I'm also not the first person to feel this way about interactions with you. Feel free to DM me if you want to discuss this.

Lets start over:

## Terminology

Forget backends, lets just use the terms **DSP** and **playback**.

* With **D**igital **S**ignal **P**rocessing I am referring to any code related to modifying audio signals in any way. Examples are
  * decoding
  * resampling
  * mixing
  * format conversion
  * effects

  This makes up the vast majority of audio related code, and I don't intend to touch any of it

* With **playback** I am referring to any code related to getting audio to, well, *play back*. This is the realm of native APIs like WASAPI and ASIO (and whatever the other platforms use), and by extension the libraries built on top of them (BASS, BASSWASAPI, BASSASIO, SDL, NAudio, ...).

## Goal

Ultimately I want to enable users to use specialized playback APIs like ASIO or WASAPI exclusive when they fulfill the respective requirements.

## Motivation

Specialized playback APIs can achieve much better latency than a generic one ever could (see my measurements further up in the thread), but aren't available (or even desirable) to everyone. I went into details, concerns, and workarounds about this during my research as well.

## Obstacles

Currently, we're using the same library (BASS) for both DSP and playback, so we need to do some decoupling first.

We already have the option to switch between playback via BASS or BASSWASAPI, and while the latter library is designed as a module of the former, we're not actually making use of their interop. This is because, instead of initializing BASSWASAPI by passing the global mixer handle along with `WASAPIPROC_BASS` to [BASS_WASAPI_Init()](https://www.un4seen.com/doc/#basswasapi/BASS_WASAPI_Init.html), we instead register a generic callback where we manually transfer the raw audio data from one to the other on a `byte[]` level. In such a configuration you technically don't even need BASS in order for BASSWASAPI to function, as it has its own device enumeration, too. So really, as far as the current code is concerned, BASS and BASSWASAPI might as well be completely independent libraries. This is also why I disagree with peppy's statement
> it's the way it's supposed to be used

because it most certainly is not. It defeats the whole point of BASSWASAPI being a BASS module. But it IS something that's needed for decoupling now.

> they have serious usability concerns

I've searched through issues, PRs, and discord chatlogs for any mention of either API, but failed to find anything of concern that isn't already addressed in this thread. The only new thing worth mentioning I found is GC stutters (brought up in the SDL PR), but since this doesn't seem to be an issue with my recent tweaks, it might have been a misattribution (even my local ASIO based build which runs at an unreasonably low 166μs buffersize works fine).

If you have any other concerns in mind, please voice them constructively.

## Design

In the end, I want each playback implementation to be nothing more than a single subclass or similar, maybe 1 extra enum constant for the settings menu. I don't have an exact design yet, I'm still at the stage of figuring out the current ecosystem after all. And if you're still worried about code complexity, just consider the proof of concept from earlier (#6649). It is really not that much.

**Comment by bdach** [comment]
> I'm really disappointed to receive such a prejudiced lecture in response to a trivial technical question. I know you probably mean no evil, but please try to be more mindful about your attitude in the future

The 'lecture' is [XY problem prevention](https://en.wikipedia.org/wiki/XY_problem) and a measure to set expectations early so that both sides are clear on where they stand. Being 'more mindful' doesn't work in my experience because if I'm not firm with where I stand the other side might not understand what my position is or think there's room for negotiation where there is none. It's not designed to be a 'comfortable' or 'encouraging' conversation.

The only other remaining move I have to not have 'attitude' and yet still convey my position is not to reply which I'm probably going to deploy going forward.

> Ultimately I want to enable users to use specialized playback APIs like ASIO or WASAPI exclusive when they fulfill the respective requirements.

Here's the thing though: I don't *personally* agree that this is something that the game should be supporting.

- WASAPI shared mode has already made marked improvements to latency. In comparison, WASAPI exclusive / ASIO don't look all that enticing. It's questionable whether the remaining gains from exclusive audio are even relevant in terms of human cognition or other sources of timing inaccuracies like frame pacing or thread scheduling.
- (Personal opinion) I don't think the game should be able to take exclusive control over device audio in any circumstances.
- If the game is going to be interfacing natively with WASAPI / ASIO, then we (read: the osu! dev team) are indefinitely on the hook for maintaining a native audio implementation forevermore, which right now is more or less handled for us by BASS (or will potentially be handled for us by SDL, if that ever happens). You've even mentioned that documentation on these is scarce and it gets to levels of near alchemy. I don't want to be beholden to maintaining this code.
- I won't even mention the support burden of this. People have so many hardware configurations and I can guarantee that more audio knobs to twist is going to result in us getting a lot more support requests for arcane configurations being broken.
   
  This is already evident by what the linux pipewire crowd is reporting to us (and we basically drop all of it on the spot because we can't fix any of it), and will get amplified by incorrect technical advice proliferating which has been the case several times over in the past and we're still fighting with (see e.g. "unlimited frame rate always"). Even shared WASAPI has begun to show signs of this after the recent release (https://github.com/ppy/osu/issues/35513, https://github.com/ppy/osu/issues/35533).

In light of the above - to *me* personally - WASAPI shared mode is the [80/20 solution](https://en.wikipedia.org/wiki/Pareto_principle) to audio latency and I don't see strong incentives to pushing any further. Maybe you'll have more luck convincing other maintainers, but right now, I haven't seen anything that would make me interested.

A more productive work effort in terms of issues with the audio subsystem would be investigating and hopefully resolving https://github.com/ppy/osu/issues/31843 (note that there have been a few reports claiming that WASAPI shared mode does fix that already).

**Comment by LastExceed** [comment]
> Being 'more mindful' doesn't work in my experience
> &
> It's not designed to be a 'comfortable' or 'encouraging' conversation

...wow. That is shockingly self-aware. My codolences for the awful world you seem to live in. Genuinely. I'm not even gonna argue the XY thing anymore, it is clear that you've already been vulnerated way more than I have. Hope you find peace some day.

---

>  It's questionable whether the remaining gains from exclusive audio are even relevant in terms of human cognition

The scepticism is understandable. However, these APIs wouldn't be so popular in the pro audio world if the difference they make wasn't meaningful. Just wait and see for yourself when I got a PR ready

> other sources of timing inaccuracies like frame pacing or thread scheduling

Neither of those is remotely comparable in magnitude, and at least one of them isn't even audio related

> I don't think the game should be able to take exclusive control over device audio in any circumstances.

I totally understand (and share) the apprehensive feeling when there's only 1 output device available, but thats not really the use-case anyway. We could even inhibit such a configuration entirely if we really wanted (though I don't see the benefit of doing so). With multiple outputs however, (which are otherwise completely dormant anyway) this is a quite trivial thing to do

> indefinitely on the hook for maintaining a native audio implementation forevermore

No you're not. Nobody is forcing you to do anything. And if things go south so badly that you really cant do nothing (zero day or whatever) then you can just do what you've already done before: `//`

> more or less handled for us by BASS
> &
> documentation on these is scarce and it gets to levels of near alchemy

and see how well that turned out. Not only did BASS not remove the need to be familiar with the underlying API, it actually made the research MORE difficult (closed source, and its own documentation isnt much better), only to reward us with literally doubling our latency in a way we cannot control.

> more support requests for arcane configurations being broken

so instead of *some* users being unable to use these, you'd rather have *nobody* being able to?

> 80 / 20

kinda sad that this is the bar for audio latency in the worlds leading rythm game

---

I'm gonna drop the thread now as well, as none of this is really moving things forward. I'm just gonna work on a PR so there's actually something to talk about, then you guys can actually try things out (or not), and whatever happens, happens.

**Comment by Devocub** [comment]
Great to see it's finally implemented! 👍
  
@LastExceed I implemented same thing in McOsu few years ago(2021) and have been using it since. Except I didn't use Event flag for BASS because it made no difference for me or maybe there's something else in code.
I posted my results in McOsu discord server and on osu! forum (for some reason screenshots are now lost from osu! servers but they are present in webarchive - 7th post) https://web.archive.org/web/20230115212434/https://osu.ppy.sh/community/forums/topics/794196?n=7 . Note these tests show full chain latency, from keyboard scanning, usb polling, game engine, BASS, audio driver, audio hardware, etc.
  
But I also implemented a feature which changes sample rate of playing hitsound based on how far it is from ideal - higher pitch for earlier, lower pitch for later. Example: [youtu.be/p96Bob8A8J4](youtu.be/p96Bob8A8J4) . This required me to use `BASS_ChannelSetAttribute(handle, BASS_ATTRIB_FREQ, 44101 for example);` in each play() call (it is applied right below this [line](https://github.com/McKay42/McEngine/blob/f2f5721409c17ce4decc20eaf2a1827ce9aafd87/McEngine/src/Engine/SoundEngine.cpp#L690) in my personal fork). After few hours of playing hitsounds (many thousands) I noticed slight visual lags, they were caused by BASS and were noticeable due to McOsu being single-threaded (up to 15 ms before another hitsound starts to be playing). This will be not noticeable in osu! and osu!lazer (or old osu! and osu! ?) because they are async. It seems something happens in BASS due to this frequency change, seems like there are a lot of dangling streams silently staying forever in playing queue? idk. [discord message where I discovered it](https://discord.com/channels/210047741097803786/222386288081371136/981762022402293831). I dirty fixed it by forcefully freeing 6 oldest streams no matter what if their amount is >12.  [in this commit](https://github.com/McKay42/McEngine/commit/e3476c1a1a533fe184685f0105ec8fb625e280fe) modify like that

```c++
if (m_danglingWasapiStreams.size() > 12) // execute only when more than 12 are active
	for (size_t i=0; i<m_danglingWasapiStreams.size()-6; i++) // remove 6 oldest streams
		{
			//if (BASS_ChannelIsActive(m_danglingWasapiStreams[i]) != BASS_ACTIVE_PLAYING) // ← line is commented. even if it's playing it will be freed
			BASS_StreamFree(m_danglingWasapiStreams[i]);
			...
		}
```
this is acceptable for casual human playing but totally not acceptable for more advanced cases like longer hitsounds or aspire or high bpm maps such as centipede - at moments of a lot of burst-playing hitsounds some of them will be terminated prematurely. This bug should be fixed in BASS library code. I recommend to thoroughly test it in osu!. Also this hitsound pitch feature is weird at first but cool and worth being implemented.
  
P.S. don't forget about this bug https://github.com/miniant-git/REAL/issues/9 . It happens with anything which asks windows for small period, and persists even after the application is not running anymore

**Comment by LastExceed** [comment]
RE: miniant-git/REAL#9
This is definitely good to be aware of. I looked into it and left a comment there.

RE: McOsu developments
Please don't take this the wrong way, but that's off-topic. While I do personally find it interesting, this is not the right place for such tangents. I'll happily discuss those lags if you DM me though!

---
While I'm back here, I wanna bring up a potential solution to the problem of exclusive audio not really being viable for single-device setups: We could capture the output of the sound server and mix it into the osu's output ourselves. This would add a second indirection to the audio streams of other applications (app -> sound server -> osu -> hardware), but for secondary application the added latency is insignificant. Combined with switching back to shared mode when the game window isn't focused (discussed further up in the thread), this should remedy most concerns about exclusive audio, save for code complexity, which I still plan to make a case against in form of a PR

**Comment by Devocub** [comment]
You seem to misunderstood me. I believe this bug is in BASS library so it should be reproducible in osu! the same way - playing many times hitsounds with BASS_ATTRIB_FREQ flag set (in shared low latency mode).
https://github.com/ppy/osu-framework/blob/master/osu.Framework/Audio/BassRelativeFrequencyHandler.cs#L73
https://github.com/ppy/osu/blob/master/osu.Game/Graphics/UserInterface/HoverClickSounds.cs#L77
I just advise to test if this bug is still present because it is nasty and appears hidden late during gaming because something in BASS accumulates. Play, for example, PlayClickSample() 60k times and use same technique to measure audio latency (without restarting audio engine anyhow obviously). Good, if it's not reproducible.

**Comment by LastExceed** [comment]
Small update on the driver situation:

I couldn't resist the temptation to try implementing WDM-KS directly, just to see what its actually like. I still think its a bad idea (even more so now than I did before), but it was quite insightful.

PortAudio is literally the only resource on the topic I can find (everything else is either closed source, or points back to PortAudio), so I once again spent an excruciating amount of time making sense of its horribly copy-pasty, hardcore memory-overoptimized source (even they themselves currently lack a maintainer that understands it, lol), and learned a lot about drivers in the process, which is why I now feel much more comfortable recommending to switch. The bottomline ultimately didn't change:

> I could imagine that it breaks stuff like OEM configuration software or propietary stream formats and whatnot

It obviously depends on the individual hardware, so just try and see if it works. Worst thing that can happen is that the device is no longer recognized as an audio endpoint, in which case you can simply switch back. This is done the same way, and pretty straight-forward.

As for progress on alternative playback APIs (i.e. actually completing this issue), it is currently blocked by #6697

**Comment by remi-scarlet-1** [comment]
If I understood correctly from the discussion in this issue there are no plans on supporting ASIO?

I've had extremely bad audio latency on the Windows build of osu!lazer (specifically with my Scarlet 3i3). In comparison, on the same machine running the Linux build of osu!lazer with pipewire the hitsound latency is near instantaneous.

I assume most osu players likely aren't running the game through a dedicated audio interface but for me and a few friends of mine it's a real blocker preventing us from making the switch to lazer. It's a shame really...

**Comment by LastExceed** [comment]
instead of "not having plans to support it", I think it is more accurate to speak of "***having*** plans to ***not*** support it"

i really wanted to make this happen. I spent so much time on this. I did the research. I wrote the proof of concept. I've literally been playing on a dev build with ASIO support for months, eagerly awaiting the day I can get it upstreamed. i did everything I could to bring this to the community. Yet my latest PR remains pending to this day, ghosted halfway through the review because I wasnt having their shit.

Maintainers have been frustrating to work with at every step of the way, and I just can't take it anymore. My love for the game is dead, and my enthusiam for the technology too got killed like never before.

You are right: This really is a shame.

**Comment by remi-scarlet-1** [comment]
Would you be willing to share your progress? Do you have a public fork with ASIO support?

I'd love to try it out even if it doesn't make it upstream. It would be an even bigger shame if we didn't get a chance to play around offline with the feature if you already spend all this time implementing it.

**Comment by LastExceed** [comment]
Its not in a state where you can just build and run without tailoring it to your specific machine first, but anything worth posting for development purposes, I have already posted.

If you're only looking for the low latency experience, [KeyASIO](https://github.com/Milkitic/KeyAsio.Net) should do the trick
